providers/implementations/keymgmt/mlx_kmgmt.c

```diff
@@ -0,0 +1,796 @@
+                       "keys cannot be mutated");
+        return 0;
+    }
+    /* An unlikely failure mode is the parameter having some unexpected type */
+    if (!OSSL_PARAM_get_octet_string_ptr(p, &pubenc, &publen))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        OPENSSL_free(key->propq);
+        key->propq = NULL;
+        if (!OSSL_PARAM_get_utf8_string(p, &key->propq, 0))
+            return 0;
+    }
+
+    if (publen != key->minfo->pubkey_bytes + key->xinfo->pubkey_bytes) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_KEY);
+        return 0;
+    }
+
+    return load_keys(key, pubenc, publen, NULL, 0);
+}
+
+static int mlx_kem_gen_set_params(void *vgctx, const OSSL_PARAM params[])
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    const OSSL_PARAM *p;
+
+    if (gctx == NULL)
+        return 0;
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_PROPERTIES);
+    if (p != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
+            return 0;
+        OPENSSL_free(gctx->propq);
+        if ((gctx->propq = OPENSSL_strdup(p->data)) == NULL)
+            return 0;
+    }
+    return 1;
+}
+
+static void *mlx_kem_gen_init(int v, OSSL_LIB_CTX *libctx, int selection,
+                              const OSSL_PARAM params[])
+{
+    PROV_ML_KEM_GEN_CTX *gctx = NULL;
+
+    /*
+     * We can only generate private keys, check that the selection is
+     * appropriate.
+     */
+    if (!ossl_prov_is_running()
+        || (selection & minimal_selection) == 0
+        || (gctx = OPENSSL_zalloc(sizeof(*gctx))) == NULL)
+        return NULL;
+
+    gctx->variant = v;
+    gctx->libctx = libctx;
+    gctx->selection = selection;
+    if (mlx_kem_gen_set_params(gctx, params))
+        return gctx;
+
+    mlx_kem_gen_cleanup(gctx);
+    return NULL;
+}
+
+static const OSSL_PARAM *mlx_kem_gen_settable_params(ossl_unused void *vgctx,
+                                                     ossl_unused void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    return settable;
+}
+
+static void *mlx_kem_gen(void *vgctx, OSSL_CALLBACK *osslcb, void *cbarg)
+{
+    PROV_ML_KEM_GEN_CTX *gctx = vgctx;
+    MLX_KEY *key;
+    char *propq = gctx->propq;
+
+    if (gctx == NULL
+        || (gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) ==
+            OSSL_KEYMGMT_SELECT_PUBLIC_KEY)
+        return NULL;
+
+    /* Lose ownership of propq */
+    gctx->propq = NULL;
+    if ((key = mlx_kem_key_new(gctx->variant, gctx->libctx, propq)) == NULL)
+        return NULL;
+
+    if ((gctx->selection & OSSL_KEYMGMT_SELECT_KEYPAIR) == 0)
+        return key;
+
+    /* For now, using the same "propq" for all components */
+    key->mkey = EVP_PKEY_Q_keygen(gctx->libctx, gctx->propq,
```

Comment:
We just set `gctx->propq` to NULL above - so this doesn't look correct.

Reply:
Well spotted. Fixed.